Post Github status for the entire mender-qa pipeline.

kacf opened the pull request:
This has several advantages compared to the existing method which
posts the status from within each job:

* If we add a new stage or job, there is no risk of forgetting to add
  a status update there.

* The status stays yellow until the whole pipeline is finished,
  instead of flipping between green and yellow as it does
  currently. This helps prevent mismerging because the pipeline looks
  green when it's actually unfinished.

* Potentially we could remove the much more complicated per-job status
  updates, and save some code, but they are still useful, and we
  already have them, so I'm not touching that now.

Signed-off-by: Kristian Amlie <kristian.amlie@northern.tech>

kacf commented:
Don't review this yet, there are some problems here.

lluiscampos commented on scripts/github_pull_request_status:
```diff
@@ -71,7 +71,7 @@ for key in $(compgen -A export); do
     pr_status_endpoint=https://api.github.com/repos/mendersoftware/$repo/statuses/$git_commit
 
-    curl -iv -H "Authorization: bearer $GITHUB_BOT_TOKEN" \
+    curl -H "Authorization: bearer $GITHUB_BOT_TOKEN" \
```
Yes, please! :clap:

lluiscampos commented on .gitlab-ci.yml:
```diff
@@ -100,6 +105,41 @@ github:failure:
+  stage: .pre
+  # Keep overhead low by using a small image with curl preinstalled.
+  image: appropriate/curl
+  before_script:
+    - curl -Lf -o $CI_PROJECT_DIR/github_pull_request_status
+      https://github.com/mendersoftware/mender-qa/raw/$CI_COMMIT_SHA/scripts/github_pull_request_status
+    - chmod ugo+x $CI_PROJECT_DIR/github_pull_request_status
```
We don't need to download it, right? This is the mender-qa pipeline so the script is already in `$CI_PROJECT_DIR/scripts/...`

lluiscampos replied:
(same for `success` and `failure` jobs)

kacf replied:
Doh! Of course! I had it in my head that this would be "reused" for other repositories, but of course it all still happens in the mender-qa pipeline. Fixing!

lluiscampos commented on .gitlab-ci.yml:
```diff
@@ -82,11 +82,16 @@ include:
     file: '.gitlab-ci-github-status-updates.yml'
```
Should we then remove these?

kacf replied:
Yes, you're right, otherwise we get double statuses. I'll remove it.

lluiscampos commented on scripts/github_pull_request_status:
```diff
@@ -39,35 +39,18 @@ for key in $(env | sed -e 's/=.*//'); do
+        git_commit="$(echo "$response" | sed -ne '/^ *"sha":/{s/^ *"sha": *"\([^"]*\)".*$/\1/; p}')"
+        if [ $? -ne 0 ]; then
+            continue
+        fi
```
Awesome idea! I remember that at the very start I added the full workspace in `backend-tests` and `integration` jobs just to be able to report to GitHub...

lluiscampos commented on scripts/github_pull_request_status:
```diff
@@ -39,35 +39,18 @@ for key in $(env | sed -e 's/=.*//'); do
+        fi
+        # Ideally we should use jq here, but this is run in so many different
+        # jobs, that we will use the fully portable sed instead.
+        git_commit="$(echo "$response" | sed -ne '/^ *"sha":/{s/^ *"sha": *"\([^"]*\)".*$/\1/; p}')"
```
I am not following on the sed script, but trust you on this...

kacf replied:
I'm actually rethinking my own decision here... I think I will try to add jq after all, it's not that many jobs to fix.

kacf commented:
Updated! Mostly just the expected small fixes, but I also switched to jq which requires installing it in the various pipelines. During this I noticed some unneeded caching and update options to apk which I cleaned up.
